File: codeowners-validator/src/main/kotlin/nl/basjes/codeowners/validator/utils/ProblemTable.java
/*
 * Yet Another UserAgent Analyzer
 * Copyright (C) 2013-2025 Niels Basjes
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.basjes.codeowners.validator.utils;

import lombok.Getter;
import org.slf4j.Logger;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.stream.Collectors;

import static nl.basjes.codeowners.validator.utils.LogColor.BLACK_BACKGROUND_BRIGHT;
import static nl.basjes.codeowners.validator.utils.LogColor.BLUE_BOLD_BRIGHT;
import static nl.basjes.codeowners.validator.utils.LogColor.BLUE_BRIGHT;
import static nl.basjes.codeowners.validator.utils.LogColor.RED_BOLD;
import static nl.basjes.codeowners.validator.utils.LogColor.RED_BRIGHT;
import static nl.basjes.codeowners.validator.utils.LogColor.RESET;
import static nl.basjes.codeowners.validator.utils.LogColor.YELLOW_BRIGHT;

@Getter
public final class ProblemTable extends StringTable{
    private final List<Problem> problems = new ArrayList<>();

    public ProblemTable() {
        super();
        withHeaders("Section", "Expression", "Approver", "Problem");
    }

    public boolean hasFatalErrors() {
        return getNumberOfFatalErrors() > 0;
    }

    public boolean hasErrors() {
        return getNumberOfErrors() > 0;
    }

    public boolean hasWarnings() {
        return getNumberOfWarnings() > 0;
    }

    public int getNumberOfProblems() {
        return problems.size();
    }

    public long getNumberOfFatalErrors() {
        return problems.stream().filter(problem -> problem instanceof Problem.Fatal).count();
    }

    public long getNumberOfErrors() {
        return problems.stream().filter(problem -> problem instanceof Problem.Error).count();
    }

    public long getNumberOfWarnings() {
        return problems.stream().filter(problem -> problem instanceof Problem.Warning).count();
    }

    public boolean contains(Problem problem) {
        return problems.contains(problem);
    }

    public boolean isEmpty() {
        return problems.isEmpty();
    }

    private void addLineInfo(StringBuilder sb, String line)    { addLine(sb, BLUE_BRIGHT,      "INFO ", RESET,            line); }
    private void addLineWarning(StringBuilder sb, String line) { addLine(sb, YELLOW_BRIGHT,    "WARN ", YELLOW_BRIGHT,    line); }
    private void addLineError(StringBuilder sb, String line)   { addLine(sb, RED_BRIGHT,       "ERROR", RED_BRIGHT,       line); }
    private void addLineFatal(StringBuilder sb, String line)   { addLine(sb, BLACK_BACKGROUND_BRIGHT.toString() + RED_BOLD,  "FATAL", BLACK_BACKGROUND_BRIGHT.toString() + RED_BOLD,  line); }
    private void addLineUnknown(StringBuilder sb, String line) { addLine(sb, BLUE_BOLD_BRIGHT, "?????", BLUE_BOLD_BRIGHT, line); }

    private void addLine(StringBuilder sb, Object tagColor, String tag, Object lineColor, String line) {
        sb.append("[").append(tagColor).append(tag).append(RESET).append("] : ") .append(lineColor).append(line).append(RESET).append('\n');
    }

    @Override
    public String toString() {
        StringBuilder buffer = new StringBuilder("\n");
        addLineInfo(buffer, writeSeparator());
        addLineInfo(buffer, writeHeaders());
        addLineInfo(buffer, writeSeparator());
        for (Problem problem : problems) {
            if (problem instanceof Problem.Info) {
                addLineInfo(buffer, writeLine(problem));
                continue;
            }
            if (problem instanceof Problem.Warning) {
                addLineWarning(buffer, writeLine(problem));
                continue;
            }
            if (problem instanceof Problem.Error) {
                addLineError(buffer, writeLine(problem));
                continue;
            }
            if (problem instanceof Problem.Fatal) {
                addLineFatal(buffer, writeLine(problem));
                continue;
            }
            if (problem != null) {
                addLineUnknown(buffer, writeLine(problem));
            }
        }
        addLineInfo(buffer, writeSeparator());

        return buffer.toString();
    }

    public void toLogAsBlock(Logger logger) {
        if (hasFatalErrors() || hasErrors()) {
            logger.error(this.toString());
            return;
        }
        if (hasWarnings()) {
            logger.warn(this.toString());
            return;
        }
        logger.info(this.toString());
    }


    public void toLog(Logger logger) {
        logger.info(writeSeparator());
        logger.info(writeHeaders());
        logger.info(writeSeparator());
        for (Problem problem : problems) {
            if (problem instanceof Problem.Info) {
                logger.info(writeLine(problem));
                continue;
            }
            if (problem instanceof Problem.Warning) {
                logger.warn(writeLine(problem));
                continue;
            }
            if (problem instanceof Problem.Error) {
                logger.error(writeLine(problem));
                continue;
            }
            if (problem instanceof Problem.Fatal) {
                logger.error(writeLine(problem));
                continue;
            }
            if (problem != null) {
                logger.info(writeLine(problem));
            }
        }
        logger.info(writeSeparator());
    }

    private String writeLine(Problem problem) {
        return writeLine(Arrays.asList(
            problem.getSection(),
            problem.getExpression(),
            problem.getApprover(),
            problem.getDescription(),
            problem.getMarker()
        ));
    }

    public void addProblem(Problem problem) {
        problems.add(problem);
        addRow(problem.getSection(),
            problem.getExpression(),
            problem.getApprover(),
            problem.getDescription(),
            problem.getMarker());
    }

    public String toProblemMessageGroupedString() {
        Map<String, Set<String>> problemMessageToUsersMap = new TreeMap<>();
        problems.forEach(problem -> problemMessageToUsersMap.computeIfAbsent(problem.getDescription(), s -> new HashSet<>()).add(problem.getApprover()));
        StringTable table = new StringTable();
        table.withHeaders("Message", "Affected users");
        problemMessageToUsersMap.forEach((message, users) -> table.addRow(message, users.stream().sorted().distinct().collect(Collectors.joining(", "))));
        return "\n"+table;
    }
}

File: codeowners-validator/src/main/kotlin/nl/basjes/codeowners/validator/utils/ProblemTable.kt
/*
 * CodeOwners Tools
 * Copyright (C) 2023-2026 Niels Basjes
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.basjes.codeowners.validator.utils

import nl.basjes.codeowners.validator.utils.LogColor.RESET
import nl.basjes.codeowners.validator.utils.Problem.Error
import nl.basjes.codeowners.validator.utils.Problem.Fatal
import nl.basjes.codeowners.validator.utils.Problem.Info
import nl.basjes.codeowners.validator.utils.Problem.ProblemColor.COLOR_INFO_TAG
import nl.basjes.codeowners.validator.utils.Problem.ProblemColor.COLOR_INFO_TEXT
import nl.basjes.codeowners.validator.utils.Problem.Tags.INFO_TAG
import nl.basjes.codeowners.validator.utils.Problem.Warning
import org.slf4j.Logger
import java.util.stream.Collectors

class ProblemTable : StringTable() {

    init {
        withHeaders("Section", "Expression", "Approver", "Problem")
    }
    private val theProblems: MutableList<Problem> = mutableListOf()
    val problems: List<Problem>
        get() = theProblems

    fun hasFatalErrors(): Boolean {
        return this.numberOfFatalErrors > 0
    }

    fun hasErrors(): Boolean {
        return this.numberOfErrors > 0
    }

    fun hasWarnings(): Boolean {
        return this.numberOfWarnings > 0
    }

    val numberOfProblems: Int
        get() = theProblems.size

    val numberOfFatalErrors: Long
        get() = theProblems.stream()
            .filter { it is Fatal }
            .count()

    val numberOfErrors: Long
        get() = theProblems.stream()
            .filter { it is Error }
            .count()

    val numberOfWarnings: Long
        get() = theProblems.stream()
            .filter { it is Warning }
            .count()

    fun contains(problem: Problem?): Boolean {
        return theProblems.contains(problem)
    }

    val isEmpty: Boolean
        get() = theProblems.isEmpty()

    override fun toString(): String {
        clearCaches();
        val buffer = StringBuilder("\n")
        fun addLineInfo(line: String) {
            buffer.append("$INFO_TAG : ${COLOR_INFO_TEXT}${line}${RESET}\n")
        }

        addLineInfo(writeSeparator())
        addLineInfo(writeHeaders())
        addLineInfo(writeSeparator())
        for (problem in theProblems) {
            buffer.append("${problem.tag} : ${writeLine(problem)}\n")
        }
        addLineInfo( writeSeparator())

        return buffer.toString()
    }

    fun toLogAsBlock(logger: Logger) {
        if (hasFatalErrors() || hasErrors()) {
            logger.error(this.toString())
            return
        }
        if (hasWarnings()) {
            logger.warn(this.toString())
            return
        }
        logger.info(this.toString())
    }


    fun toLog(logger: Logger) {
        clearCaches();
        logger.info(writeSeparator())
        logger.info(writeHeaders())
        logger.info(writeSeparator())
        for (problem in theProblems) {
            when(problem) {
                is Info     -> logger.info(writeLine(problem))
                is Warning  -> logger.warn(writeLine(problem))
                is Error    -> logger.error(writeLine(problem))
                is Fatal    -> logger.error(writeLine(problem))
            }
        }
        logger.info(writeSeparator())
    }

    private fun writeLine(problem: Problem) =
        "${problem.messageColor}${writeLine(
            listOf(
                problem.section,
                problem.expression,
                problem.approver,
                problem.description,
                problem.marker,
            )
            )}${RESET}"

    fun addProblem(problem: Problem) {
        theProblems.add(problem)
        addRow(
            problem.section,
            problem.expression,
            problem.approver,
            problem.description,
            problem.marker
        )
    }

    fun toProblemMessageGroupedString(): String {
        val problemMessageToUsersMap: MutableMap<String, MutableSet<String>> = mutableMapOf()
        theProblems.forEach {
            problemMessageToUsersMap
                .computeIfAbsent(it.description) { mutableSetOf() }
                .add(it.approver)
        }
        val table = StringTable()
        table.withHeaders("Message", "Affected users")
        problemMessageToUsersMap
            .toSortedMap()
            .forEach { (message, users) ->
            table.addRow(
                message,
                users.stream().sorted().distinct().collect(Collectors.joining(", "))
            )
        }
        return "\n" + table
    }
}
